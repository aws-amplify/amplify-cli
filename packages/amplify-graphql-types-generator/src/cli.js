#!/usr/bin/env node

import * as glob from 'glob';
import * as process from 'process';
import * as path from 'path';
import * as yargs from 'yargs';

import { downloadSchema, introspectSchema, printSchema, generate } from '.';
import { ToolError, logError } from './errors'

import 'source-map-support/register'

// Make sure unhandled errors in async code are propagated correctly
process.on('unhandledRejection', (error) => { throw error });

process.on('uncaughtException', handleError);

function handleError(error) {
  logError(error);
  process.exit(1);
}

yargs
  .command(
    '$0 [input..]',
    'Generate code from a GraphQL schema and query documents',
    {
      schema: {
        demand: true,
        describe: 'Path to GraphQL schema file. (Defaults to using .graphqlconfig or schema.json)',
        normalize: true,
        coerce: path.resolve,
      },
      output: {
        describe: 'Output directory for the generated files',
        normalize: true,
        coerce: path.resolve,
      },
      target: {
        demand: false,
        describe: 'Code generation target language',
        choices: ['swift', 'scala', 'json', 'ts', 'typescript', 'flow', 'flow-modern', 'angular'],
        default: 'angular'
      },
      only: {
        describe: 'Parse all input files, but only output generated code for the specified file [Swift only]',
        normalize: true,
        coerce: path.resolve,
      },
      namespace: {
        demand: false,
        describe: 'Optional namespace for generated types [currently Swift and Scala-only]',
        type: 'string'
      },
      "passthrough-custom-scalars": {
        demand: false,
        describe: "Don't attempt to map custom scalars [temporary option]",
        default: false
      },
      "custom-scalars-prefix": {
        demand: false,
        describe: "Prefix for custom scalars. (Implies that passthrough-custom-scalars is true if set)",
        default: '',
        normalize: true
      },
      "add-typename": {
        demand: false,
        describe: "For non-swift targets, always add the __typename GraphQL introspection type when generating target types",
        default: false
      },
      "use-flow-exact-objects": {
        demand: false,
        describe: "Use Flow exact objects for generated types [flow-modern only]",
        default: false,
        type: 'boolean'
      },
      "tag-name": {
        demand: false,
        describe: "Name of the template literal tag used to identify template literals containing GraphQL queries in Javascript/Typescript code",
        default: 'gql'
      },
      "project-name": {
        demand: false,
        describe: "Name of the project to use in a multi-project .graphqlconfig file",
      },
      "merge-in-fields-from-fragment-spreads": {
        demand: false,
        describe: "Merge fragment fields onto its enclosing type",
        default: true,
        type: 'boolean'
      },
      "complex-object-support": {
        demand: false,
        describe: "Adds S3 wrapper code to the output. [Swift only]",
        default: 'auto',
        choices: ['yes', 'no', 'auto'],
      }
    },
    argv => {
      let { input } = argv;

      // Use glob if the user's shell was unable to expand the pattern
      if (input.length === 1 && glob.hasMagic(input[0])) {
        input = glob.sync(input[0]);
      }

      const inputPaths = input
        .map(input => path.resolve(input))
        // Sort to normalize different glob expansions between different terminals.
        .sort();

      const options = {
        passthroughCustomScalars: argv["passthrough-custom-scalars"] || argv["custom-scalars-prefix"] !== '',
        customScalarsPrefix: argv["custom-scalars-prefix"] || '',
        addTypename: argv["add-typename"],
        namespace: argv.namespace,
        mergeInFieldsFromFragmentSpreads: argv["merge-in-fields-from-fragment-spreads"],
        useFlowExactObjects: argv['use-flow-exact-objects'],
        complexObjectSupport: argv["complex-object-support"],
      };

      generate(inputPaths, argv.schema, argv.output, argv.only, argv.target, argv.tagName, options);
    },
  )
  .fail(function(message, error) {
    handleError(error ? error : new ToolError(message));
  })
  .help()
  .version()
  .strict()
  .argv
