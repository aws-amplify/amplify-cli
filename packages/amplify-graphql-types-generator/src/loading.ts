import * as fs from 'fs'

import {
  buildClientSchema,
  Source,
  concatAST,
  parse,
  DocumentNode,
  GraphQLSchema
} from 'graphql';

import { ToolError } from './errors'

export function loadSchema(schemaPath: string): GraphQLSchema {
  if (!fs.existsSync(schemaPath)) {
    throw new ToolError(`Cannot find GraphQL schema file: ${schemaPath}`);
  }
  const schemaData = require(schemaPath);

  if (!schemaData.data && !schemaData.__schema) {
    throw new ToolError('GraphQL schema file should contain a valid GraphQL introspection query result');
  }
  return buildClientSchema((schemaData.data) ? schemaData.data : schemaData);
}

function extractDocumentFromJavascript(content: string, tagName: string = 'gql'): string | null {
  const re = new RegExp(tagName + '\\s*`([^`/]*)`', 'g');

  let match
  const matches = []

  while(match = re.exec(content)) {
    const doc = match[1]
      .replace(/\${[^}]*}/g, '')

    matches.push(doc)
  }

  const doc = matches.join('\n')
  return doc.length ? doc : null;
}

export function loadAndMergeQueryDocuments(inputPaths: string[], tagName: string = 'gql'): DocumentNode {
  const sources = inputPaths.map(inputPath => {
    const body = fs.readFileSync(inputPath, 'utf8');
    if (!body) {
      return null;
    }

    if (inputPath.endsWith('.jsx') || inputPath.endsWith('.js')
      || inputPath.endsWith('.tsx') || inputPath.endsWith('.ts')
    ) {
      const doc = extractDocumentFromJavascript(body.toString(), tagName);
      return doc ? new Source(doc, inputPath) : null;
    }

    return new Source(body, inputPath);
  }).filter(source => source);

  return concatAST((sources as Source[]).map(source => parse(source)));
}
