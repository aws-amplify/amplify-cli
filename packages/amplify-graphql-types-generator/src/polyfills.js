import 'core-js/fn/object/values';
import 'core-js/fn/object/entries';
import 'core-js/fn/array/flat-map';
