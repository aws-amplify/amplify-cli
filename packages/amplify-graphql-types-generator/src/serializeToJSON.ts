import {
  isType,
  GraphQLType,
  GraphQLScalarType,
  GraphQLEnumType,
  GraphQLInputObjectType,
} from 'graphql';

import { LegacyCompilerContext } from './compiler/legacyIR';

export default function serializeToJSON(context: LegacyCompilerContext) {
  return serializeAST({
    operations: Object.values(context.operations),
    fragments: Object.values(context.fragments),
    typesUsed: context.typesUsed.map(serializeType),
  }, '\t');
}

export function serializeAST(ast: any, space?: string) {
  return JSON.stringify(ast, function(_, value) {
    if (isType(value)) {
      return String(value);
    } else {
      return value;
    }
  }, space);
}

function serializeType(type: GraphQLType) {
  if (type instanceof GraphQLEnumType) {
    return serializeEnumType(type);
  } else if (type instanceof GraphQLInputObjectType) {
    return serializeInputObjectType(type);
  } else if (type instanceof GraphQLScalarType) {
    return serializeScalarType(type);
  } else {
    throw new Error(`Unexpected GraphQL type: ${type}`);
  }
}

function serializeEnumType(type: GraphQLEnumType) {
  const { name, description } = type;
  const values = type.getValues();

  return {
    kind: 'EnumType',
    name,
    description,
    values: values.map(value => (
      {
        name: value.name,
        description: value.description,
        isDeprecated: value.isDeprecated,
        deprecationReason: value.deprecationReason
      }
    ))
  }
}

function serializeInputObjectType(type: GraphQLInputObjectType) {
  const { name, description } = type;
  const fields = Object.values(type.getFields());

  return {
    kind: 'InputObjectType',
    name,
    description,
    fields: fields.map(field => ({
      name: field.name,
      type: String(field.type),
      description: field.description,
      defaultValue: field.defaultValue
    }))
  }
}

function serializeScalarType(type: GraphQLScalarType) {
  const { name, description } = type;

  return {
    kind: 'ScalarType',
    name,
    description
  }
}
