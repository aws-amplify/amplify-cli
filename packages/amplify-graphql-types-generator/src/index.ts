import './polyfills';

export { default as generate } from './generate';
