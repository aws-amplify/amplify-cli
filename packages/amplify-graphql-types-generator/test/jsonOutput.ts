import { GraphQLSchema, buildSchema, parse } from 'graphql';
import { compileToLegacyIR } from '../src/compiler/legacyIR';
import serializeToJSON from '../src/serializeToJSON';

import { loadSchema } from '../src/loading';
const starWarsSchema = loadSchema(require.resolve('./fixtures/starwars/schema.json'));

function compileFromSource(source: string, schema: GraphQLSchema = starWarsSchema) {
  const document = parse(source);
  return compileToLegacyIR(schema, document, {
    mergeInFieldsFromFragmentSpreads: false,
    addTypename: true,
  });
}

describe('JSON output', function() {
  test(`should generate JSON output for a query with an enum variable`, function() {
    const context = compileFromSource(`
      query HeroName($episode: Episode) {
        hero(episode: $episode) {
          name
        }
      }
    `);

    const output = serializeToJSON(context);

    expect(output).toMatchSnapshot();
  });

  test(`should generate JSON output for a query with a nested selection set`, function() {
    const context = compileFromSource(`
      query HeroAndFriendsNames {
        hero {
          name
          friends {
            name
          }
        }
      }
    `);

    const output = serializeToJSON(context);

    expect(output).toMatchSnapshot();
  });

  test(`should generate JSON output for a query with a fragment spread and nested inline fragments`, function() {
    const context = compileFromSource(`
      query HeroAndDetails {
        hero {
          id
          ...CharacterDetails
        }
      }

      fragment CharacterDetails on Character {
        name
        ... on Droid {
          primaryFunction
        }
        ... on Human {
          height
        }
      }
    `);

    const output = serializeToJSON(context);

    expect(output).toMatchSnapshot();
  });

  test(`should generate JSON output for a mutation with an enum and an input object variable`, function() {
    const context = compileFromSource(`
      mutation CreateReview($episode: Episode, $review: ReviewInput) {
        createReview(episode: $episode, review: $review) {
          stars
          commentary
        }
      }
    `);

    const output = serializeToJSON(context);

    expect(output).toMatchSnapshot();
  });

  test(`should generate JSON output for an input object type with default field values`, function() {
    const schema = buildSchema(`
      type Query {
        someField(input: ComplexInput!): String!
      }

      input ComplexInput {
        string: String = "Hello"
        customScalar: Date = "2017-04-16"
        listOfString: [String] = ["test1", "test2", "test3"]
        listOfInt: [Int] = [1, 2, 3]
        listOfEnums: [Episode] = [JEDI, EMPIRE]
        listOfCustomScalar: [Date] = ["2017-04-16", "2017-04-17", "2017-04-18"]
      }

      scalar Date

      enum Episode {
        NEWHOPE
        EMPIRE
        JEDI
      }
    `);

    const context = compileFromSource(
      `
      query QueryWithComplexInput($input: ComplexInput) {
        someField(input: $input)
      }
      `,
      schema
    );

    const output = serializeToJSON(context);

    expect(output).toMatchSnapshot();
  });
});
