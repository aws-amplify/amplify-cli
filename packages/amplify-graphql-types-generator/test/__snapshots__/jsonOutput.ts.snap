// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON output should generate JSON output for a mutation with an enum and an input object variable 1`] = `
"{
	\\"operations\\": [
		{
			\\"filePath\\": \\"GraphQL request\\",
			\\"operationName\\": \\"CreateReview\\",
			\\"operationType\\": \\"mutation\\",
			\\"rootType\\": \\"Mutation\\",
			\\"variables\\": [
				{
					\\"name\\": \\"episode\\",
					\\"type\\": \\"Episode\\"
				},
				{
					\\"name\\": \\"review\\",
					\\"type\\": \\"ReviewInput\\"
				}
			],
			\\"source\\": \\"mutation CreateReview($episode: Episode, $review: ReviewInput) {\\\\n  createReview(episode: $episode, review: $review) {\\\\n    __typename\\\\n    stars\\\\n    commentary\\\\n  }\\\\n}\\",
			\\"fields\\": [
				{
					\\"responseName\\": \\"createReview\\",
					\\"fieldName\\": \\"createReview\\",
					\\"type\\": \\"Review\\",
					\\"args\\": [
						{
							\\"name\\": \\"episode\\",
							\\"value\\": {
								\\"kind\\": \\"Variable\\",
								\\"variableName\\": \\"episode\\"
							}
						},
						{
							\\"name\\": \\"review\\",
							\\"value\\": {
								\\"kind\\": \\"Variable\\",
								\\"variableName\\": \\"review\\"
							}
						}
					],
					\\"isConditional\\": false,
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null,
					\\"fields\\": [
						{
							\\"responseName\\": \\"__typename\\",
							\\"fieldName\\": \\"__typename\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false
						},
						{
							\\"responseName\\": \\"stars\\",
							\\"fieldName\\": \\"stars\\",
							\\"type\\": \\"Int!\\",
							\\"isConditional\\": false,
							\\"description\\": \\"The number of stars this review gave, 1-5\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						},
						{
							\\"responseName\\": \\"commentary\\",
							\\"fieldName\\": \\"commentary\\",
							\\"type\\": \\"String\\",
							\\"isConditional\\": false,
							\\"description\\": \\"Comment about the movie\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						}
					],
					\\"fragmentSpreads\\": [],
					\\"inlineFragments\\": []
				}
			],
			\\"fragmentSpreads\\": [],
			\\"inlineFragments\\": [],
			\\"fragmentsReferenced\\": [],
			\\"sourceWithFragments\\": \\"mutation CreateReview($episode: Episode, $review: ReviewInput) {\\\\n  createReview(episode: $episode, review: $review) {\\\\n    __typename\\\\n    stars\\\\n    commentary\\\\n  }\\\\n}\\",
			\\"operationId\\": \\"0ce790562a5f5a8afc1e0d002d0436768d7649e4ec8b4aa816d3849f66768c7e\\"
		}
	],
	\\"fragments\\": [],
	\\"typesUsed\\": [
		{
			\\"kind\\": \\"EnumType\\",
			\\"name\\": \\"Episode\\",
			\\"description\\": \\"The episodes in the Star Wars trilogy\\",
			\\"values\\": [
				{
					\\"name\\": \\"NEWHOPE\\",
					\\"description\\": \\"Star Wars Episode IV: A New Hope, released in 1977.\\",
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null
				},
				{
					\\"name\\": \\"EMPIRE\\",
					\\"description\\": \\"Star Wars Episode V: The Empire Strikes Back, released in 1980.\\",
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null
				},
				{
					\\"name\\": \\"JEDI\\",
					\\"description\\": \\"Star Wars Episode VI: Return of the Jedi, released in 1983.\\",
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null
				}
			]
		},
		{
			\\"kind\\": \\"InputObjectType\\",
			\\"name\\": \\"ReviewInput\\",
			\\"description\\": \\"The input object sent when someone is creating a new review\\",
			\\"fields\\": [
				{
					\\"name\\": \\"stars\\",
					\\"description\\": \\"0-5 stars\\",
					\\"type\\": \\"Int!\\"
				},
				{
					\\"name\\": \\"commentary\\",
					\\"description\\": \\"Comment about the movie, optional\\",
					\\"type\\": \\"String\\"
				},
				{
					\\"name\\": \\"favorite_color\\",
					\\"description\\": \\"Favorite color, optional\\",
					\\"type\\": \\"ColorInput\\"
				}
			]
		},
		{
			\\"kind\\": \\"InputObjectType\\",
			\\"name\\": \\"ColorInput\\",
			\\"description\\": \\"The input object sent when passing in a color\\",
			\\"fields\\": [
				{
					\\"name\\": \\"red\\",
					\\"description\\": \\"\\",
					\\"type\\": \\"Int!\\"
				},
				{
					\\"name\\": \\"green\\",
					\\"description\\": \\"\\",
					\\"type\\": \\"Int!\\"
				},
				{
					\\"name\\": \\"blue\\",
					\\"description\\": \\"\\",
					\\"type\\": \\"Int!\\"
				}
			]
		}
	]
}"
`;

exports[`JSON output should generate JSON output for a query with a fragment spread and nested inline fragments 1`] = `
"{
	\\"operations\\": [
		{
			\\"filePath\\": \\"GraphQL request\\",
			\\"operationName\\": \\"HeroAndDetails\\",
			\\"operationType\\": \\"query\\",
			\\"rootType\\": \\"Query\\",
			\\"variables\\": [],
			\\"source\\": \\"query HeroAndDetails {\\\\n  hero {\\\\n    __typename\\\\n    id\\\\n    ...CharacterDetails\\\\n  }\\\\n}\\",
			\\"fields\\": [
				{
					\\"responseName\\": \\"hero\\",
					\\"fieldName\\": \\"hero\\",
					\\"type\\": \\"Character\\",
					\\"isConditional\\": false,
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null,
					\\"fields\\": [
						{
							\\"responseName\\": \\"__typename\\",
							\\"fieldName\\": \\"__typename\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false
						},
						{
							\\"responseName\\": \\"id\\",
							\\"fieldName\\": \\"id\\",
							\\"type\\": \\"ID!\\",
							\\"isConditional\\": false,
							\\"description\\": \\"The ID of the character\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						}
					],
					\\"fragmentSpreads\\": [
						\\"CharacterDetails\\"
					],
					\\"inlineFragments\\": []
				}
			],
			\\"fragmentSpreads\\": [],
			\\"inlineFragments\\": [],
			\\"fragmentsReferenced\\": [
				\\"CharacterDetails\\"
			],
			\\"sourceWithFragments\\": \\"query HeroAndDetails {\\\\n  hero {\\\\n    __typename\\\\n    id\\\\n    ...CharacterDetails\\\\n  }\\\\n}\\\\nfragment CharacterDetails on Character {\\\\n  __typename\\\\n  name\\\\n  ... on Droid {\\\\n    primaryFunction\\\\n  }\\\\n  ... on Human {\\\\n    height\\\\n  }\\\\n}\\",
			\\"operationId\\": \\"9ea2d583a2bcf98ddcedc95a807e34011ff3c0f66fabaf01fa0169e4a4159121\\"
		}
	],
	\\"fragments\\": [
		{
			\\"typeCondition\\": \\"Character\\",
			\\"possibleTypes\\": [
				\\"Human\\",
				\\"Droid\\"
			],
			\\"fragmentName\\": \\"CharacterDetails\\",
			\\"filePath\\": \\"GraphQL request\\",
			\\"source\\": \\"fragment CharacterDetails on Character {\\\\n  __typename\\\\n  name\\\\n  ... on Droid {\\\\n    primaryFunction\\\\n  }\\\\n  ... on Human {\\\\n    height\\\\n  }\\\\n}\\",
			\\"fields\\": [
				{
					\\"responseName\\": \\"__typename\\",
					\\"fieldName\\": \\"__typename\\",
					\\"type\\": \\"String!\\",
					\\"isConditional\\": false
				},
				{
					\\"responseName\\": \\"name\\",
					\\"fieldName\\": \\"name\\",
					\\"type\\": \\"String!\\",
					\\"isConditional\\": false,
					\\"description\\": \\"The name of the character\\",
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null
				}
			],
			\\"fragmentSpreads\\": [],
			\\"inlineFragments\\": [
				{
					\\"typeCondition\\": \\"Droid\\",
					\\"possibleTypes\\": [
						\\"Droid\\"
					],
					\\"fields\\": [
						{
							\\"responseName\\": \\"__typename\\",
							\\"fieldName\\": \\"__typename\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false
						},
						{
							\\"responseName\\": \\"name\\",
							\\"fieldName\\": \\"name\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false,
							\\"description\\": \\"What others call this droid\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						},
						{
							\\"responseName\\": \\"primaryFunction\\",
							\\"fieldName\\": \\"primaryFunction\\",
							\\"type\\": \\"String\\",
							\\"isConditional\\": false,
							\\"description\\": \\"This droid's primary function\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						}
					],
					\\"fragmentSpreads\\": []
				},
				{
					\\"typeCondition\\": \\"Human\\",
					\\"possibleTypes\\": [
						\\"Human\\"
					],
					\\"fields\\": [
						{
							\\"responseName\\": \\"__typename\\",
							\\"fieldName\\": \\"__typename\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false
						},
						{
							\\"responseName\\": \\"name\\",
							\\"fieldName\\": \\"name\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false,
							\\"description\\": \\"What this human calls themselves\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						},
						{
							\\"responseName\\": \\"height\\",
							\\"fieldName\\": \\"height\\",
							\\"type\\": \\"Float\\",
							\\"isConditional\\": false,
							\\"description\\": \\"Height in the preferred unit, default is meters\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						}
					],
					\\"fragmentSpreads\\": []
				}
			]
		}
	],
	\\"typesUsed\\": []
}"
`;

exports[`JSON output should generate JSON output for a query with a nested selection set 1`] = `
"{
	\\"operations\\": [
		{
			\\"filePath\\": \\"GraphQL request\\",
			\\"operationName\\": \\"HeroAndFriendsNames\\",
			\\"operationType\\": \\"query\\",
			\\"rootType\\": \\"Query\\",
			\\"variables\\": [],
			\\"source\\": \\"query HeroAndFriendsNames {\\\\n  hero {\\\\n    __typename\\\\n    name\\\\n    friends {\\\\n      __typename\\\\n      name\\\\n    }\\\\n  }\\\\n}\\",
			\\"fields\\": [
				{
					\\"responseName\\": \\"hero\\",
					\\"fieldName\\": \\"hero\\",
					\\"type\\": \\"Character\\",
					\\"isConditional\\": false,
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null,
					\\"fields\\": [
						{
							\\"responseName\\": \\"__typename\\",
							\\"fieldName\\": \\"__typename\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false
						},
						{
							\\"responseName\\": \\"name\\",
							\\"fieldName\\": \\"name\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false,
							\\"description\\": \\"The name of the character\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						},
						{
							\\"responseName\\": \\"friends\\",
							\\"fieldName\\": \\"friends\\",
							\\"type\\": \\"[Character]\\",
							\\"isConditional\\": false,
							\\"description\\": \\"The friends of the character, or an empty list if they have none\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null,
							\\"fields\\": [
								{
									\\"responseName\\": \\"__typename\\",
									\\"fieldName\\": \\"__typename\\",
									\\"type\\": \\"String!\\",
									\\"isConditional\\": false
								},
								{
									\\"responseName\\": \\"name\\",
									\\"fieldName\\": \\"name\\",
									\\"type\\": \\"String!\\",
									\\"isConditional\\": false,
									\\"description\\": \\"The name of the character\\",
									\\"isDeprecated\\": false,
									\\"deprecationReason\\": null
								}
							],
							\\"fragmentSpreads\\": [],
							\\"inlineFragments\\": []
						}
					],
					\\"fragmentSpreads\\": [],
					\\"inlineFragments\\": []
				}
			],
			\\"fragmentSpreads\\": [],
			\\"inlineFragments\\": [],
			\\"fragmentsReferenced\\": [],
			\\"sourceWithFragments\\": \\"query HeroAndFriendsNames {\\\\n  hero {\\\\n    __typename\\\\n    name\\\\n    friends {\\\\n      __typename\\\\n      name\\\\n    }\\\\n  }\\\\n}\\",
			\\"operationId\\": \\"d424c7c31d5147b382972b3c316ace3d51f2f039b53852a280d707df2848f715\\"
		}
	],
	\\"fragments\\": [],
	\\"typesUsed\\": []
}"
`;

exports[`JSON output should generate JSON output for a query with an enum variable 1`] = `
"{
	\\"operations\\": [
		{
			\\"filePath\\": \\"GraphQL request\\",
			\\"operationName\\": \\"HeroName\\",
			\\"operationType\\": \\"query\\",
			\\"rootType\\": \\"Query\\",
			\\"variables\\": [
				{
					\\"name\\": \\"episode\\",
					\\"type\\": \\"Episode\\"
				}
			],
			\\"source\\": \\"query HeroName($episode: Episode) {\\\\n  hero(episode: $episode) {\\\\n    __typename\\\\n    name\\\\n  }\\\\n}\\",
			\\"fields\\": [
				{
					\\"responseName\\": \\"hero\\",
					\\"fieldName\\": \\"hero\\",
					\\"type\\": \\"Character\\",
					\\"args\\": [
						{
							\\"name\\": \\"episode\\",
							\\"value\\": {
								\\"kind\\": \\"Variable\\",
								\\"variableName\\": \\"episode\\"
							}
						}
					],
					\\"isConditional\\": false,
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null,
					\\"fields\\": [
						{
							\\"responseName\\": \\"__typename\\",
							\\"fieldName\\": \\"__typename\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false
						},
						{
							\\"responseName\\": \\"name\\",
							\\"fieldName\\": \\"name\\",
							\\"type\\": \\"String!\\",
							\\"isConditional\\": false,
							\\"description\\": \\"The name of the character\\",
							\\"isDeprecated\\": false,
							\\"deprecationReason\\": null
						}
					],
					\\"fragmentSpreads\\": [],
					\\"inlineFragments\\": []
				}
			],
			\\"fragmentSpreads\\": [],
			\\"inlineFragments\\": [],
			\\"fragmentsReferenced\\": [],
			\\"sourceWithFragments\\": \\"query HeroName($episode: Episode) {\\\\n  hero(episode: $episode) {\\\\n    __typename\\\\n    name\\\\n  }\\\\n}\\",
			\\"operationId\\": \\"f6e76545cd03aa21368d9969cb39447f6e836a16717823281803778e7805d671\\"
		}
	],
	\\"fragments\\": [],
	\\"typesUsed\\": [
		{
			\\"kind\\": \\"EnumType\\",
			\\"name\\": \\"Episode\\",
			\\"description\\": \\"The episodes in the Star Wars trilogy\\",
			\\"values\\": [
				{
					\\"name\\": \\"NEWHOPE\\",
					\\"description\\": \\"Star Wars Episode IV: A New Hope, released in 1977.\\",
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null
				},
				{
					\\"name\\": \\"EMPIRE\\",
					\\"description\\": \\"Star Wars Episode V: The Empire Strikes Back, released in 1980.\\",
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null
				},
				{
					\\"name\\": \\"JEDI\\",
					\\"description\\": \\"Star Wars Episode VI: Return of the Jedi, released in 1983.\\",
					\\"isDeprecated\\": false,
					\\"deprecationReason\\": null
				}
			]
		}
	]
}"
`;

exports[`JSON output should generate JSON output for an input object type with default field values 1`] = `
"{
	\\"operations\\": [
		{
			\\"filePath\\": \\"GraphQL request\\",
			\\"operationName\\": \\"QueryWithComplexInput\\",
			\\"operationType\\": \\"query\\",
			\\"rootType\\": \\"Query\\",
			\\"variables\\": [
				{
					\\"name\\": \\"input\\",
					\\"type\\": \\"ComplexInput\\"
				}
			],
			\\"source\\": \\"query QueryWithComplexInput($input: ComplexInput) {\\\\n  someField(input: $input)\\\\n}\\",
			\\"fields\\": [
				{
					\\"responseName\\": \\"someField\\",
					\\"fieldName\\": \\"someField\\",
					\\"type\\": \\"String!\\",
					\\"args\\": [
						{
							\\"name\\": \\"input\\",
							\\"value\\": {
								\\"kind\\": \\"Variable\\",
								\\"variableName\\": \\"input\\"
							},
							\\"type\\": \\"ComplexInput!\\"
						}
					],
					\\"isConditional\\": false,
					\\"isDeprecated\\": false
				}
			],
			\\"fragmentSpreads\\": [],
			\\"inlineFragments\\": [],
			\\"fragmentsReferenced\\": [],
			\\"sourceWithFragments\\": \\"query QueryWithComplexInput($input: ComplexInput) {\\\\n  someField(input: $input)\\\\n}\\",
			\\"operationId\\": \\"55e208ec5def01b99b65c64b666d06e95558a6e455c52eb5c2307c4775afb805\\"
		}
	],
	\\"fragments\\": [],
	\\"typesUsed\\": [
		{
			\\"kind\\": \\"InputObjectType\\",
			\\"name\\": \\"ComplexInput\\",
			\\"fields\\": [
				{
					\\"name\\": \\"string\\",
					\\"type\\": \\"String\\",
					\\"defaultValue\\": \\"Hello\\"
				},
				{
					\\"name\\": \\"customScalar\\",
					\\"type\\": \\"Date\\",
					\\"defaultValue\\": \\"2017-04-16\\"
				},
				{
					\\"name\\": \\"listOfString\\",
					\\"type\\": \\"[String]\\",
					\\"defaultValue\\": [
						\\"test1\\",
						\\"test2\\",
						\\"test3\\"
					]
				},
				{
					\\"name\\": \\"listOfInt\\",
					\\"type\\": \\"[Int]\\",
					\\"defaultValue\\": [
						1,
						2,
						3
					]
				},
				{
					\\"name\\": \\"listOfEnums\\",
					\\"type\\": \\"[Episode]\\",
					\\"defaultValue\\": [
						\\"JEDI\\",
						\\"EMPIRE\\"
					]
				},
				{
					\\"name\\": \\"listOfCustomScalar\\",
					\\"type\\": \\"[Date]\\",
					\\"defaultValue\\": [
						\\"2017-04-16\\",
						\\"2017-04-17\\",
						\\"2017-04-18\\"
					]
				}
			]
		},
		{
			\\"kind\\": \\"ScalarType\\",
			\\"name\\": \\"Date\\"
		},
		{
			\\"kind\\": \\"EnumType\\",
			\\"name\\": \\"Episode\\",
			\\"values\\": [
				{
					\\"name\\": \\"NEWHOPE\\",
					\\"isDeprecated\\": false
				},
				{
					\\"name\\": \\"EMPIRE\\",
					\\"isDeprecated\\": false
				},
				{
					\\"name\\": \\"JEDI\\",
					\\"isDeprecated\\": false
				}
			]
		}
	]
}"
`;
